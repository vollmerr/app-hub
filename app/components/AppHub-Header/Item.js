import styled from 'styled-components';

import theme from 'utils/theme';

const focusStyle = (props) => `
  background: ${props.dark ? theme.blueDark : theme.themePrimary};
  border-bottom: 2px solid ${props.dark ? theme.orange : theme.orangeLighter};
`;

const Item = styled.div`
  display: flex;
  align-items: center;
  height: ${theme.hub.headerHeight};
  border-bottom: 2px solid transparent;

  ${(props) => props.isLink &&
    `&:hover,
    &:active,
    &:focus {
      ${focusStyle(props)}
    }`
  }
  ${(props) => props.checked &&
    focusStyle(props)
  }
`;

export default Item;
