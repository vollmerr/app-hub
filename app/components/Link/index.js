/**
*
* Link
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import { Link as RouterLink } from 'react-router-dom';
import styled from 'styled-components';

export const StyledLink = styled(RouterLink) `
  text-decoration: none;
`;

export const A = styled.a`
  text-decoration: none;
`;

class Link extends React.PureComponent {
  render() {
    const { to, href, children, ...props } = this.props;

    if (to) {
      return (
        <StyledLink to={to} {...props}>
          {children}
        </StyledLink>
      );
    }

    return (
      <A href={href} target={'_blank'} rel={'noopener noreferrer'} {...props}>
        {children}
      </A>
    );
  }
}

Link.propTypes = {
  to: PropTypes.string,
  href: PropTypes.string,
  children: PropTypes.node,
};

export default Link;
