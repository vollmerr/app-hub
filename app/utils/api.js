import { delay } from 'redux-saga';
import { call, put } from 'redux-saga/effects';
import decode from 'jwt-decode';
import 'whatwg-fetch';

import { authUserDone } from '../containers/AppHub/actions';


export const DEFAULT_EXPIRE = 1000;
export const TOKEN = 'id_token';


/**
 * Determines if should fetch based off a cache time
 *
 * @param {Date} date             - date of cache
 * @param {Number} cacheSeconds   - number of seconds to cache
 *
 * @return {Bool}                 - if should fetch
 */
export function shouldFetch(date, cacheSeconds = 350) {
  return !date || (new Date(date).getTime() - (cacheSeconds * 1000)) > new Date().getTime();
}


/**
 * Sets a token for authenication in localStorage
 *
 * @param {String} token  - token to set
 */
export function setToken(token) {
  localStorage.setItem(TOKEN, token);
}


/**
 * Gets the authenication token from localStorage
 * @return {String}     - authenication token
 */
export function getToken() {
  return localStorage.getItem(TOKEN);
}


/**
 * Clears the authenication token from localStorage
 */
export function clearToken() {
  localStorage.removeItem(TOKEN);
}


/**
 * Helper function to valdate that token exists and is not expired
 *
 * @param {object} token  - jwt token
 *
 * @return {bool}         - valid ? true : false
 */
export function validToken(token) {
  try {
    const { exp } = decode(token);
    const now = new Date().getTime().toString().substring(0, 10);

    return Number(exp) > Number(now);
  } catch (e) {
    return false;
  }
}


/**
 * Helper function to put jwt token into global redux store
 *
 * @param {String} token    - jwt token
 *
 * @return {Number}         - exipration time
 */
export function* putToken(token) {
  try {
    const {
      sub: sam,
      exp,
      iat,
      ...rest
    } = decode(token);

    const expire = (exp - iat) * 1000; // convert from ms to s

    yield put(authUserDone({ sam, expire, ...rest }));
    return expire;
  } catch (error) {
    throw new Error(error);
  }
}


/**
 * Authenticates a user
 * Checks local storage for token first
 *
 * @param {Object} action   - redux action to dispatch (must have 'type')
 * @param {Number} maxTries - max tries to re attempt authenicating on fail
 */
export function* authenticate(action, maxTries = 3) {
  let expire = DEFAULT_EXPIRE; // try again in 1s by default
  let tries = 0;
  let error = new Error('Failed to authenticate');
  while (tries < maxTries) { // eslint-disable-line
    const token = getToken();
    // if valid local token use that, otherwise get new token
    if (validToken(token)) {
      // update user in global appHub state
      expire = yield call(putToken, token);
      // reset max tries
      tries = 0;
    } else {
      try {
        const options = {
          method: 'get',
          credentials: 'include',
        };
        const { id_token } = yield call(request, API.JWT, options); // eslint-disable-line
        // set into local storage for future authentication caching
        setToken(id_token);
        // update user in global appHub state
        expire = yield call(putToken, id_token);
        // reset max tries
        tries = 0;
      } catch (err) {
        // increment max tries
        tries += 1;
        error = err;
      }
      // max number of tires, set error
      if (tries === maxTries) {
        yield put(authUserDone(error));
      }
    }
    // authenicate again when token has expired
    yield call(delay, expire);
  }
}


/**
 * Parses the JSON returned by a network request
 *
 * @param  {object} response A response from a network request
 *
 * @return {object}          The parsed JSON from the request
 */
export function parseJSON(response) {
  if (response.status === 204 || response.status === 205) {
    return null;
  }
  return response.json();
}


/**
 * Checks if a network request came back fine, and throws an error if not
 *
 * @param  {object} response   A response from a network request
 *
 * @return {object|undefined} Returns either the response, or throws an error
 */
export function checkStatus(response) {
  if (response.status >= 200 && response.status < 300) {
    return response;
  }

  const error = new Error(response.statusText);
  error.response = response;
  throw error;
}


/**
 * Requests a URL, returning a promise
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export function request(url, options) {
  return fetch(url, options)
    .then(checkStatus)
    .then(parseJSON);
}


/**
 * Requests a URL with jwt token in header for authorization, returning a promise
 * Waits until jwt token is set in localStorage before making call
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export function requestWithToken(url, options = {}) {
  let timer = null;

  const requestIfToken = () => { // eslint-disable-line
    // get users jwt token form storage else they dont have one
    const token = localStorage.getItem('id_token') || null;
    // valid token, call api
    if (validToken(token)) {
      clearTimeout(timer);
      // config settings to send api
      const newOptions = {
        method: options.method || 'get',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
        ...options,
      };
      // append optional body to request (POSTs)
      if (options.body) {
        newOptions.body = JSON.stringify(options.body);
      }

      return request(url, newOptions);
    }
    // wait again until token set
    timer = setTimeout(requestIfToken, 100);
  };

  return requestIfToken();
}
