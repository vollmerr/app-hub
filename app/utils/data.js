/**
 * Formats an item object to the correct format
 * for rendering
 *
 * @param {array} item      - item to format
 * @param {string} name     - internal field name for lookup
 * @param {object} field    - field format info
 * @param {object} enums    - display names for enums
 *
 * @return {string}         - formatted item
 */
export function formatItem(item, name, field, enums = {}) {
  let content = item[name];
  // convert all data to array for mapping over
  if (!Array.isArray(content)) {
    content = [content];
  }
  // go through each item, mapping to the correct format
  content = content.map((x) => {
    // is enum mapping
    if (enums[name]) {
      return enums[name][x];
    }
    // is date
    if (field.data && field.data.type === 'DATE') {
      return isNaN(Date.parse(x)) ? '' : new Date(x).toISOString().substr(0, 10);
    }
    return x;
  });
  // join them back as string seperated by commas
  return content.join(', ');
}

/**
 * Formats an array of item objects to the correct format
 * for rendering
 *
 * @param {array} items     - items to format
 * @param {object} fields   - lookup table of fields
 * @param {object} enums    - display names for enums
 *
 * @return {array}          - formatted items as objects of strings
 */
export function formatItems(items, fields, enums) {
  const formattedItems = [];

  items.forEach((item, i) => {
    formattedItems[i] = {};
    Object.keys(item).forEach((k) => {
      formattedItems[i][k] = formatItem(item, k, fields[k], enums);
    });
  });

  return formattedItems;
}

/**
 * Maps a object to the format Office-UI-Fabric-React DetailsList`s expect
 *
 * @param {object} obj    - object to map
 * @param {array} include - keys to include, if empty includes all
 * @param {array} exclude - keys to exclude
 *
 * @return {object}       - mapped object to use in List
 */
export function mapToColumns(obj, include = [], exclude = []) {
  const keys = include.length ?
    include :
    Object.keys(obj);

  return keys
    .filter((key) => !exclude.includes(key))
    .map((key) => ({
      key: obj[key].name,
      name: obj[key].label,
      fieldName: obj[key].name,
      isResizable: true,
      ariaLabel: obj[key].ariaLabel,
      minWidth: obj[key].minWidth || 100,
      maxWidth: obj[key].maxWidth || 200,
      data: obj[key].data || {},
    }));
}
