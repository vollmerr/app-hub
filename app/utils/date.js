export const dateWithoutTime = (date) => {
  if (date && !isNaN(new Date(date))) {
    const d = new Date(date);
    d.setHours(0, 0, 0, 0);
    return d;
  }
  return undefined;
};
