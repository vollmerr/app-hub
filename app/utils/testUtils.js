import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

/**
 * Tests a styled-component for basic tests
 * such as rending correctly and adopting attributes
 * @param {Object} Component    - React component under test
 * @param {Object} extend       - React component the component under test extends /  inherits from
 */
export function testStyledComponent(Component, extend = null) {
  describe(`<${Component.displayName} />`, () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<Component />);
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it(`should render an <${Component.target}> tag`, () => {
      expect(wrapper.type()).toEqual(Component.target);
    });

    it('should have a className attribute', () => {
      expect(wrapper.prop('className')).toBeDefined();
    });

    it('should adopt a valid attribute', () => {
      const id = 'test';
      wrapper.setProps({ id });
      expect(wrapper.prop('id')).toEqual(id);
    });

    if (extend) {
      it(`should extend ${extend.name}`, () => {
        expect(wrapper.find(extend).exists()).toEqual(true);
      });
    } else {
      it('should not adopt an invalid attribute', () => {
        const attribute = 'test';
        wrapper.setProps({ attribute });
        expect(wrapper.prop('attribute')).toBeUndefined();
      });
    }
  });
}

/**
 * Tests a components for basic mapDispatchToProps information
 * @param {func} mapDispatchToProps   - components mapDispatchToProps function
 * @param {object} actions            - functions in mapDispatchToProps to test
 */
export function testMapDispatchToProps(mapDispatchToProps, actions) {
  const onAction = (action) => (
    `on${action.charAt(0).toUpperCase()}${action.slice(1)}`
  );

  let dispatch;
  let dispatchProps;
  beforeEach(() => {
    dispatch = jest.fn();
    dispatchProps = mapDispatchToProps(dispatch);
  });

  Object.keys(actions).forEach((action) => {
    describe(onAction(action), () => {
      it('should be injected', () => {
        expect(dispatchProps[onAction(action)]).toBeDefined();
      });

      it('should dispatch when called', () => {
        dispatchProps[onAction(action)]();
        expect(dispatch).toHaveBeenCalled();
      });
    });
  });
}
