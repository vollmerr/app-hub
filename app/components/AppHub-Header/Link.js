import React from 'react';
import styled from 'styled-components';
import { CommandButton, ElementType } from 'office-ui-fabric-react/lib/Button';
import { Link as RouterLink } from 'react-router-dom';

import theme from 'utils/theme';
import PropTypes from 'prop-types';
// import HeaderItem from 'components/Header/HeaderItem';
import Item from './Item';

const LinkStyled = styled(RouterLink) `
  text-decoration: none;
`;

const A = LinkStyled.withComponent('a');

const ButtonStyled = styled(CommandButton) `
  color: ${theme.themePrimary};
  display: block;
  vertical-align: middle;
  border-width: 0;
  margin: 0;
  padding: 0 15px;
  min-width: ${theme.hub.headerHeight};
  height: ${theme.hub.headerHeight};
  box-sizing: border-box;
  text-align: center;
  line-height: normal;
  overflow: visible;
  transition: background-color .467s cubic-bezier(.1,.9,.2,1) 34ms;
  pointer-events: all;
  font-size: ${(props) => props.size || '28px'};
`;

class Link extends React.PureComponent {
  render() {
    const { panel, title, iconProps, onClick, to, href, text, children } = this.props;

    const isLink = onClick || to || href;

    const Button = () => (
      <ButtonStyled
        onClick={onClick ? () => onClick(panel) : null}
        iconProps={iconProps}
        title={title}
        ariaLabel={title}
      >
        {text}
        {children}
      </ButtonStyled>
    );

    // handle internal routing
    if (to) {
      return (
        <Item isLink={isLink}>
          <LinkStyled to={to} >
            <Button />
          </LinkStyled>
        </Item>
      );
    } else if (href) {
      return (
        <Item isLink={isLink}>
          <A href={href} >
            <Button />
          </A>
        </Item>
      );
    }

    // no links
    return (
      <Item isLink={isLink}>
        <Button />
      </Item>
    );
  }
}

Link.propTypes = {
  panel: PropTypes.string,
  title: PropTypes.string,
  iconProps: PropTypes.object,
  onClick: PropTypes.func,
  text: PropTypes.string,
  to: PropTypes.string,
  href: PropTypes.string,
  children: PropTypes.node,
};

export default Link;
