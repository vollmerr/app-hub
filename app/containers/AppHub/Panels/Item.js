import React from 'react';
import styled from 'styled-components';

import Link from '../../../components/Link';
import theme from '../../../utils/theme';


export const Wrapper = styled.li`
  list-style-type: none;
  margin: 0 -20px;
  &:hover {
    background: ${theme.neutralLighter};
  }
`;


export const StlyedLink = styled(Link) `
  color: ${theme.white};
  display: block;
  padding: 15px;
  &:hover,
  &:active,
  &:focus {
    color: ${theme.neutralDark};
  }
`;


class Item extends React.PureComponent {
  render() {
    return (
      <Wrapper>
        <StlyedLink {...this.props} />
      </Wrapper>
    );
  }
}


export default Item;
