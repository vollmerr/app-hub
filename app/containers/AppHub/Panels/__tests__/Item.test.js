import React from 'react';
import { shallow } from 'enzyme';

import { testStyledComponent } from 'utils/testUtils';

import Link from 'components/Link';
import Item, { Wrapper, StlyedLink } from '../Item';


testStyledComponent(Wrapper);
testStyledComponent(StlyedLink, Link);


describe('<Item />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Item />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should contain a link', () => {
    expect(wrapper.find(StlyedLink).length).toEqual(1);
  });

  it('should pass props to the link', () => {
    const to = '/test-route';
    wrapper.setProps({ to });
    expect(wrapper.find(StlyedLink).prop('to')).toEqual(to);
  });
});
