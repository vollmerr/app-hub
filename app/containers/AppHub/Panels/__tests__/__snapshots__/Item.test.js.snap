// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> should render correctly 1`] = `
<Item__Wrapper>
  <Item__StlyedLink />
</Item__Wrapper>
`;

exports[`<Item__StlyedLink /> should render correctly 1`] = `
.c0 {
  color: #ffffff;
  display: block;
  padding: 15px;
}

.c0:hover,
.c0:active,
.c0:focus {
  color: #212121;
}

<Link
  className="c0"
/>
`;

exports[`<Item__Wrapper /> should render correctly 1`] = `
.c0 {
  list-style-type: none;
  margin: 0 -20px;
}

.c0:hover {
  background: #f4f4f4;
}

<li
  className="c0"
/>
`;
